### ISSUE TITLE
Devise throwing strange exception when under heavy load

### ISSUE THREAD

**Issue by Siraris** [plataformatec/devise#4346]
So this is a very strange issue.  Currently, I'm trying to load test our app in anticipation of heavy traffic next week, and it seems that Devise could be the culprit of our failing load tests... or it could be a red herring.

For all intents and purposes, our setup should be able to handle in excess of 500 requests a second, yet any attempt to reproduce over 20 requests a second results in this strange error below.  The odd thing is, it only seems to happen on certain routes of our site.  If I run the test on one of our more database intensive pages, the error below is thrown.  If I run the same test on a less intensive page (for example, our health check URL which does a simple request to our database and displays the result) this error is not generated.  To contrast, on the more page that makes more requests to the database, I can barely get 20 requests per second through to the server before Puma starts logging this error, while when hitting the health check URL, I can get 200+ requests per second with no issue from Rails or Puma.

The other odd thing is this - the actions I'm hitting both live in the same controller, and both call the same initializations methods on application_controller before being run.  set_raven_context is the first before_action method called in application_controller, so I would assume the test on the healthcheck URL would produce the same result - this does not seem to be the case.

```
Error during failsafe response: Devise could not find the `Warden::Proxy` instance on your request environment.
Make sure that your application is loading Devise and Warden as expected and that the `Warden::Manager` middleware is present in your middleware stack.
If you are seeing this on one of your tests, ensure that your tests are either executing the Rails middleware stack or that your tests are using the `Devise::Test::ControllerHelpers` module to inject the `request.env['warden']` object for you.
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/devise-4.2.0/lib/devise/controllers/helpers.rb:141:in `warden'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/devise-4.2.0/lib/devise/controllers/helpers.rb:124:in `current_user'
  /var/app/current/app/controllers/application_controller.rb:62:in `current_user'
  /var/app/current/app/controllers/application_controller.rb:70:in `user_signed_in?'
  /var/app/current/app/controllers/application_controller.rb:120:in `set_raven_context'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:432:in `block in make_lambda'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:164:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:164:in `block in halting'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:504:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:504:in `block in call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:504:in `each'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:504:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:92:in `__run_callbacks__'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:778:in `_run_process_action_callbacks'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:81:in `run_callbacks'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/abstract_controller/callbacks.rb:19:in `process_action'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_controller/metal/rescue.rb:29:in `process_action'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_controller/metal/instrumentation.rb:32:in `block in process_action'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/notifications.rb:164:in `block in instrument'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/notifications.rb:164:in `instrument'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_controller/metal/instrumentation.rb:30:in `process_action'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_controller/metal/params_wrapper.rb:250:in `process_action'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activerecord-4.2.7.1/lib/active_record/railties/controller_runtime.rb:18:in `process_action'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/abstract_controller/base.rb:137:in `process'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionview-4.2.7.1/lib/action_view/rendering.rb:30:in `process'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/rack-mini-profiler-0.10.1/lib/mini_profiler/profiling_methods.rb:76:in `block in profile_method'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_controller/metal.rb:196:in `dispatch'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_controller/metal/rack_delegation.rb:13:in `dispatch'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_controller/metal.rb:237:in `block in action'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/routing/route_set.rb:74:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/routing/route_set.rb:74:in `dispatch'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/routing/route_set.rb:43:in `serve'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/journey/router.rb:43:in `block in serve'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/journey/router.rb:30:in `each'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/journey/router.rb:30:in `serve'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/routing/route_set.rb:817:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/middleware/show_exceptions.rb:47:in `render_exception'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/middleware/show_exceptions.rb:35:in `rescue in call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/newrelic_rpm-3.17.0.325/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/railties-4.2.7.1/lib/rails/rack/logger.rb:38:in `call_app'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/railties-4.2.7.1/lib/rails/rack/logger.rb:20:in `block in call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/tagged_logging.rb:68:in `block in tagged'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/tagged_logging.rb:26:in `tagged'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/tagged_logging.rb:68:in `tagged'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/railties-4.2.7.1/lib/rails/rack/logger.rb:20:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/newrelic_rpm-3.17.0.325/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/middleware/request_id.rb:21:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/newrelic_rpm-3.17.0.325/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/methodoverride.rb:22:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/newrelic_rpm-3.17.0.325/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/runtime.rb:18:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/newrelic_rpm-3.17.0.325/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/activesupport-4.2.7.1/lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/newrelic_rpm-3.17.0.325/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/sendfile.rb:113:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/newrelic_rpm-3.17.0.325/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/sentry-raven-1.2.3/lib/raven/integrations/rack.rb:50:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/newrelic_rpm-3.17.0.325/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/rack-mini-profiler-0.10.1/lib/mini_profiler/profiler.rb:171:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/newrelic_rpm-3.17.0.325/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/railties-4.2.7.1/lib/rails/engine.rb:518:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/railties-4.2.7.1/lib/rails/application.rb:165:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/newrelic_rpm-3.17.0.325/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/rack-reverse-proxy-0.11.0/lib/rack_reverse_proxy/roundtrip.rb:19:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/rack-reverse-proxy-0.11.0/lib/rack_reverse_proxy/middleware.rb:25:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/newrelic_rpm-3.17.0.325/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/puma-2.14.0/lib/puma/configuration.rb:78:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/puma-2.14.0/lib/puma/server.rb:541:in `handle_request'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/puma-2.14.0/lib/puma/server.rb:388:in `process_client'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/puma-2.14.0/lib/puma/server.rb:270:in `block in run'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/puma-2.14.0/lib/puma/thread_pool.rb:106:in `call'
  /opt/rubies/ruby-2.2.5/lib/ruby/gems/2.2.0/gems/puma-2.14.0/lib/puma/thread_pool.rb:106:in `block in spawn_thread'
```

**Comment by ethirajsrinivasan** [comment]
@siraris have u injected `request.env['warden'] in ur tests

**Comment by Siraris** [comment]
@ethirajsrinivasan This is not while running tests, this is in production.

**Comment by ethirajsrinivasan** [comment]
@siraris.  Yes I can understand that it happens in production.. but can you replicate it  in some way or the other in ur local machine?

**Comment by Siraris** [comment]
@ethirajsrinivasan The only way I can replicate it is if I use something like apache benchmark against my local machine.

**Comment by bf4** [comment]
@Siraris @ethirajsrinivasan The offending lines are:

```
  /var/app/current/app/controllers/application_controller.rb:62:in `current_user'
  /var/app/current/app/controllers/application_controller.rb:70:in `user_signed_in?'
  /var/app/current/app/controllers/application_controller.rb:120:in `set_raven_context'
```

The app is using `config.exceptions_app = self.routes` and the exceptions app is inheriting from `ApplicationController`, which means it needs to `skip_callback :set_raven_context` and anything else that might look for session info.

I'd recommend not inheriting from `ActionController:Base` at all for an exceptions app.

For comparison, the default in Rails (`railties/lib/rails/application/default_middleware_stack.rb `) is

```
   def show_exceptions_app
          config.exceptions_app || ActionDispatch::PublicExceptions.new(Rails.public_path)
```

i.e. [PublicExceptions](https://github.com/rails/rails/blob/5-0-stable/actionpack/lib/action_dispatch/middleware/public_exceptions.rb) which just has a `call` method.

See https://github.com/rails/rails/blob/5-0-stable/actionpack/lib/action_controller/base.rb#L183-L191 for help on making your own minimal exceptions controller that doesn't get all the `ActiveSupport.on_load(:action_controller)` mixin effects

**Comment by ccamargov** [comment]
Hi,

Could you solve it ?, I have the same problem, when deploying the application in development environment everything works perfect, but when I try to deploy the application in production environment I get this error and I do not know why ... Any ideas?

**Comment by krainboltgreene** [comment]
@ccamargov Just to update, you need to solve it by making your exception application not inherit from `ActionController`. That means it needs to take `env` and return a Rack middleware shape:

```
[StatusCode, Headers, Body | Array<Body>]`
```

where those types are defined as:

```
KeyType = string
StatusCode = number
Headers = {
  [key: KeyType]: string | number
}
Body = string
```

tl;dr: Your "exception application" (aka middleware) has *too much knowledge*, which includes devise stuff. Delete that thing and make a function that returns the above shape.

**Comment by tegon** [comment]
I'm closing this since it looks like it is not a `devise` problem.

**Comment by eveevans** [comment]
Just for the sake of this kind of debugging, 
When you are using production Environment, is helpful to display (temporarily) Rails errors  (instead of `public/500.html` or custom app page). 

In your environment (on this case `config/environment/production.rb`) change 
`config.consider_all_requests_local  = true`

**Comment by krainboltgreene** [comment]
Absolutely do not do this.
